File: app/example/find_path.py
# coding=utf-8
# author: Xiguang Liu<g10guang@foxmail.com>
# 2018-04-29 13:18
# 题目描述：https://www.nowcoder.com/practice/b736e784e3e34731af99065031301bca?tpId=13&tqId=11177&tPage=1&rp=1&ru=%2Fta%2Fcoding-interviews&qru=%2Fta%2Fcoding-interviews%2Fquestion-ranking


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def __init__(self):
        # 记录结果的二维链表
        self.result = []
        self.curPath = []
        self.curSum = 0

    # 返回二维列表，内部每个列表表示找到的路径
    def FindPath(self, root, expectNumber):
        """
        因为结点的数值有可能为负数，所以必须遍历所有的叶子结点，也就是不能够提前判断当前你路径失败
        """
        if root is None:
            return self.result
        self.curPath.append(root.val)
        self.curSum += root.val
        # 判断是否是叶子结点，如果是叶子结点需要判断是否 curSum == expectNumber
        if root.left is None and root.right is None:
            if self.curSum == expectNumber:
                self.result.append(self.curPath[:])
        else:
            self.FindPath(root.left, expectNumber)
            self.FindPath(root.right, expectNumber)
        self.curSum -= root.val
        self.curPath.pop()
        return self.result

File: app/example/find_path.go
package main

// type TreeNode struct {
//   Val int
//   Left *TreeNode
//   Right *TreeNode
// }

/**
 * 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
 *
 * @param root TreeNode类
 * @param expectNumber int整型
 * @return int整型二维数组
 */
// https://www.nowcoder.com/practice/b736e784e3e34731af99065031301bca?tpId=13&tqId=11177&tPage=1&rp=1&ru=%2Fta%2Fcoding-interviews&qru=%2Fta%2Fcoding-interviews%2Fquestion-ranking
func FindPath(root *TreeNode, expectNumber int) [][]int {
	// 深度优先遍历二叉树
	var result [][]int
	var fn func(node *TreeNode, path []int, sum int)
	fn = func(node *TreeNode, path []int, sum int) {
		if node == nil {
			return
		}
		if node.Left == nil && node.Right == nil {
			if sum+node.Val == expectNumber {
				result = append(result, append(path, node.Val))
				return
			}
		}
		fn(node.Left, append(path, node.Val), sum+node.Val)
		fn(node.Right, append(path, node.Val), sum+node.Val)
	}
	fn(root, nil, 0)
	return result
}
